File: packages/shared-types/src/runtime.ts
import type {
  BottomPanelTab,
  ConnectionGroupMode,
  RightDrawerView,
  SidebarPane,
  UiActivity,
} from './app'
import type { ConnectionProfile, DatastoreEngine, DatastoreFamily } from './connection'
import type { AdapterCapability, ExecutionCapabilities } from './capabilities'
import type { GuardrailDecision } from './security'
import type {
  ExecutionResultEnvelope,
  ExplorerNode,
  QueryBuilderKind,
  QueryBuilderState,
  QueryLanguage,
  ScopedQueryTarget,
  QueryTabState,
  ResultPageInfo,
  ResultPayload,
  ResultRenderer,
  StructureEdge,
  StructureGroup,
  StructureMetric,
  StructureNode,
} from './workspace'

export type QueryExecutionMode = 'full' | 'selection' | 'explain'

export const DATASTORE_OPERATION_RISKS = [
  'read',
  'diagnostic',
  'write',
  'destructive',
  'costly',
] as const

export type DatastoreOperationRisk = (typeof DATASTORE_OPERATION_RISKS)[number]

export const DATASTORE_OPERATION_SCOPES = [
  'connection',
  'cluster',
  'database',
  'schema',
  'table',
  'collection',
  'document',
  'key',
  'item',
  'index',
  'query',
  'user',
  'role',
  'file',
] as const

export type DatastoreOperationScope = (typeof DATASTORE_OPERATION_SCOPES)[number]

export type DatastoreOperationExecutionSupport = 'live' | 'plan-only' | 'unsupported'

export interface DatastoreOperationManifest {
  id: string
  engine: DatastoreEngine
  family: DatastoreFamily
  label: string
  scope: DatastoreOperationScope
  risk: DatastoreOperationRisk
  requiredCapabilities: AdapterCapability[]
  supportedRenderers: ResultRenderer[]
  description: string
  requiresConfirmation: boolean
  executionSupport: DatastoreOperationExecutionSupport
  disabledReason?: string
  previewOnly?: boolean
}

export interface OperationPlan {
  operationId: string
  engine: DatastoreEngine
  summary: string
  generatedRequest: string
  requestLanguage: QueryLanguage
  destructive: boolean
  estimatedCost?: string
  estimatedScanImpact?: string
  requiredPermissions: string[]
  confirmationText?: string
  warnings: string[]
}

export interface PermissionInspection {
  engine: DatastoreEngine
  principal?: string
  effectiveRoles: string[]
  effectivePrivileges: string[]
  iamSignals: string[]
  unavailableActions: Array<{
    operationId: string
    reason: string
  }>
  warnings: string[]
}

export interface AdapterDiagnostics {
  engine: DatastoreEngine
  plans: ResultPayload[]
  profiles: ResultPayload[]
  metrics: ResultPayload[]
  queryHistory: ResultPayload[]
  costEstimates: ResultPayload[]
  warnings: string[]
}

export interface DatastoreExperienceObjectKind {
  kind: string
  label: string
  description: string
  childKinds: string[]
  queryable: boolean
  supportsContextMenu: boolean
}

export interface DatastoreExperienceAction {
  id: string
  label: string
  scope: DatastoreOperationScope
  risk: DatastoreOperationRisk
  operationId?: string
  requiresSelection: boolean
  description: string
}

export interface DatastoreExperienceBuilder {
  kind: QueryBuilderKind | 'sql-select' | 'search-dsl' | 'dynamodb-key-condition' | 'cql-partition'
  label: string
  scope: DatastoreOperationScope
  defaultMode: 'visual' | 'split' | 'raw'
}

export interface DatastoreEditableScope {
  scope: DatastoreOperationScope
  label: string
  editKinds: DataEditKind[]
  requiresPrimaryKey: boolean
  liveExecution: boolean
}

export interface DatastoreDiagnosticsTab {
  id: string
  label: string
  description: string
  defaultRenderer: ResultRenderer
}

export interface DatastoreExperienceManifest {
  engine: DatastoreEngine
  family: DatastoreFamily
  label: string
  maturity: 'mvp' | 'beta' | 'planned'
  objectKinds: DatastoreExperienceObjectKind[]
  contextActions: DatastoreExperienceAction[]
  queryBuilders: DatastoreExperienceBuilder[]
  editableScopes: DatastoreEditableScope[]
  diagnosticsTabs: DatastoreDiagnosticsTab[]
  resultRenderers: ResultRenderer[]
  safetyRules: string[]
}

export interface DatastoreExperienceResponse {
  experiences: DatastoreExperienceManifest[]
}

export type DataEditKind =
  | 'insert-row'
  | 'update-row'
  | 'delete-row'
  | 'set-field'
  | 'unset-field'
  | 'rename-field'
  | 'change-field-type'
  | 'set-key-value'
  | 'set-ttl'
  | 'delete-key'
  | 'put-item'
  | 'update-item'
  | 'delete-item'
  | 'index-document'
  | 'update-document'
  | 'delete-document'

export interface DataEditTarget {
  objectKind: string
  path: string[]
  schema?: string
  table?: string
  collection?: string
  key?: string
  documentId?: unknown
  itemKey?: Record<string, unknown>
  primaryKey?: Record<string, unknown>
}

export interface DataEditChange {
  field?: string
  path?: string[]
  value?: unknown
  valueType?: string
  newName?: string
}

export interface DataEditPlanRequest {
  connectionId: string
  environmentId: string
  editKind: DataEditKind
  target: DataEditTarget
  changes: DataEditChange[]
}

export interface DataEditPlanResponse {
  connectionId: string
  environmentId: string
  editKind: DataEditKind
  executionSupport: DatastoreOperationExecutionSupport
  plan: OperationPlan
}

export interface DataEditExecutionRequest extends DataEditPlanRequest {
  confirmationText?: string
}

export interface DataEditExecutionResponse {
  connectionId: string
  environmentId: string
  editKind: DataEditKind
  executionSupport: DatastoreOperationExecutionSupport
  executed: boolean
  plan: OperationPlan
  messages: string[]
  warnings: string[]
  result?: ExecutionResultEnvelope
  metadata?: unknown
}

export interface OperationManifestRequest {
  connectionId: string
  environmentId: string
  scope?: string
}

export interface OperationManifestResponse {
  connectionId: string
  environmentId: string
  engine: DatastoreEngine
  operations: DatastoreOperationManifest[]
}

export interface OperationPlanRequest {
  connectionId: string
  environmentId: string
  operationId: string
  objectName?: string
  parameters?: Record<string, unknown>
}

export interface OperationPlanResponse {
  connectionId: string
  environmentId: string
  plan: OperationPlan
}

export interface OperationExecutionRequest {
  connectionId: string
  environmentId: string
  operationId: string
  objectName?: string
  parameters?: Record<string, unknown>
  confirmationText?: string
  rowLimit?: number
  tabId?: string
}

export interface OperationExecutionResponse {
  connectionId: string
  environmentId: string
  operationId: string
  executionSupport: DatastoreOperationExecutionSupport
  executed: boolean
  plan: OperationPlan
  result?: ExecutionResultEnvelope
  permissionInspection?: PermissionInspection
  diagnostics?: AdapterDiagnostics
  metadata?: unknown
  messages: string[]
  warnings: string[]
}

export interface PermissionInspectionRequest {
  connectionId: string
  environmentId: string
}

export interface PermissionInspectionResponse {
  connectionId: string
  environmentId: string
  inspection: PermissionInspection
}

export interface AdapterDiagnosticsRequest {
  connectionId: string
  environmentId: string
  scope?: string
}

export interface AdapterDiagnosticsResponse {
  connectionId: string
  environmentId: string
  diagnostics: AdapterDiagnostics
}

export interface ConnectionTestRequest {
  profile: ConnectionProfile
  environmentId: string
  secret?: string
}

export interface ConnectionTestResult {
  ok: boolean
  engine: DatastoreEngine
  message: string
  warnings: string[]
  resolvedHost: string
  resolvedDatabase?: string
  durationMs?: number
}

export interface ExplorerRequest {
  connectionId: string
  environmentId: string
  limit?: number
  scope?: string
}

export interface ExplorerResponse {
  connectionId: string
  environmentId: string
  scope?: string
  summary: string
  capabilities: ExecutionCapabilities
  nodes: ExplorerNode[]
}

export interface ExplorerInspectRequest {
  connectionId: string
  environmentId: string
  nodeId: string
}

export interface ExplorerInspectResponse {
  nodeId: string
  summary: string
  queryTemplate?: string
  payload?: unknown
}

export interface StructureRequest {
  connectionId: string
  environmentId: string
  limit?: number
  scope?: string
}

export interface StructureResponse {
  connectionId: string
  environmentId: string
  engine: DatastoreEngine
  summary: string
  groups: StructureGroup[]
  nodes: StructureNode[]
  edges: StructureEdge[]
  metrics: StructureMetric[]
  truncated?: boolean
  nextCursor?: string
}

export interface ExecutionRequest {
  executionId?: string
  tabId: string
  connectionId: string
  environmentId: string
  language: QueryLanguage
  queryText: string
  selectedText?: string
  mode?: QueryExecutionMode
  rowLimit?: number
  confirmedGuardrailId?: string
}

export interface ExecutionResponse {
  executionId: string
  tab: QueryTabState
  result?: ExecutionResultEnvelope
  guardrail: GuardrailDecision
  diagnostics: string[]
}

export interface ResultPageRequest {
  tabId: string
  connectionId: string
  environmentId: string
  language: QueryLanguage
  queryText: string
  selectedText?: string
  renderer: string
  pageSize?: number
  pageIndex?: number
  cursor?: string
}

export interface ResultPageResponse {
  tabId: string
  resultId?: string
  payload: ResultPayload
  pageInfo: ResultPageInfo
  notices: string[]
}

export interface CancelExecutionRequest {
  executionId: string
  tabId?: string
}

export interface CancelExecutionResult {
  ok: boolean
  supported: boolean
  message: string
}

export interface QueryTabReorderRequest {
  orderedTabIds: string[]
}

export interface CreateScopedQueryTabRequest {
  connectionId: string
  environmentId?: string
  target: ScopedQueryTarget
}

export interface UpdateQueryBuilderStateRequest {
  tabId: string
  builderState: QueryBuilderState
  queryText?: string
}

export type LocalDatabaseCreateMode = 'empty' | 'starter'

export interface LocalDatabasePickRequest {
  engine: DatastoreEngine
  purpose: 'open' | 'create'
  currentPath?: string
}

export interface LocalDatabasePickResult {
  canceled: boolean
  path?: string
}

export interface LocalDatabaseCreateRequest {
  engine: DatastoreEngine
  path: string
  mode: LocalDatabaseCreateMode
  connectionId?: string
  environmentId?: string
}

export interface LocalDatabaseCreateResult {
  engine: DatastoreEngine
  path: string
  message: string
  warnings: string[]
}

export interface UpdateUiStateRequest {
  activeEnvironmentId?: string
  activeActivity?: UiActivity
  sidebarCollapsed?: boolean
  activeSidebarPane?: SidebarPane
  sidebarWidth?: number
  explorerFilter?: string
  explorerView?: 'tree' | 'structure'
  connectionGroupMode?: ConnectionGroupMode
  sidebarSectionStates?: Record<string, boolean>
  bottomPanelVisible?: boolean
  activeBottomPanelTab?: BottomPanelTab
  bottomPanelHeight?: number
  rightDrawer?: RightDrawerView
  rightDrawerWidth?: number
}

File: apps/desktop/src-tauri/src/adapters/runtime.rs
use super::registry::adapter_for_engine;
use super::*;

pub async fn test_connection(
    connection: &ResolvedConnectionProfile,
    warnings: Vec<String>,
) -> Result<ConnectionTestResult, CommandError> {
    let adapter = adapter_for_engine(&connection.engine)?;
    let mut result = adapter.test_connection(connection).await?;
    result.warnings.extend(warnings);
    Ok(result)
}

pub async fn list_explorer_nodes(
    connection: &ResolvedConnectionProfile,
    request: &ExplorerRequest,
) -> Result<ExplorerResponse, CommandError> {
    adapter_for_engine(&connection.engine)?
        .list_explorer_nodes(connection, request)
        .await
}

pub async fn inspect_explorer_node(
    connection: &ResolvedConnectionProfile,
    request: &ExplorerInspectRequest,
) -> Result<ExplorerInspectResponse, CommandError> {
    adapter_for_engine(&connection.engine)?
        .inspect_explorer_node(connection, request)
        .await
}

pub async fn load_structure_map(
    connection: &ResolvedConnectionProfile,
    request: &StructureRequest,
) -> Result<StructureResponse, CommandError> {
    adapter_for_engine(&connection.engine)?
        .load_structure_map(connection, request)
        .await
}

pub async fn execute(
    connection: &ResolvedConnectionProfile,
    request: &ExecutionRequest,
    notices: Vec<QueryExecutionNotice>,
) -> Result<ExecutionResultEnvelope, CommandError> {
    adapter_for_engine(&connection.engine)?
        .execute(connection, request, notices)
        .await
}

pub async fn fetch_result_page(
    connection: &ResolvedConnectionProfile,
    request: &ResultPageRequest,
) -> Result<ResultPageResponse, CommandError> {
    adapter_for_engine(&connection.engine)?
        .fetch_result_page(connection, request)
        .await
}

pub async fn cancel(
    connection: &ResolvedConnectionProfile,
    request: &CancelExecutionRequest,
) -> Result<CancelExecutionResult, CommandError> {
    adapter_for_engine(&connection.engine)?
        .cancel(connection, request)
        .await
}

pub fn experience_manifests() -> Vec<DatastoreExperienceManifest> {
    super::registry::manifests()
        .into_iter()
        .map(|manifest| experience_manifest_for_manifest(&manifest))
        .collect()
}

pub fn operation_manifests(
    connection: &ResolvedConnectionProfile,
) -> Result<Vec<DatastoreOperationManifest>, CommandError> {
    Ok(adapter_for_engine(&connection.engine)?.operation_manifests())
}

pub async fn plan_operation(
    connection: &ResolvedConnectionProfile,
    operation_id: &str,
    object_name: Option<&str>,
    parameters: Option<&BTreeMap<String, Value>>,
) -> Result<OperationPlan, CommandError> {
    adapter_for_engine(&connection.engine)?
        .plan_operation(connection, operation_id, object_name, parameters)
        .await
}

pub async fn execute_operation(
    connection: &ResolvedConnectionProfile,
    request: &OperationExecutionRequest,
) -> Result<OperationExecutionResponse, CommandError> {
    adapter_for_engine(&connection.engine)?
        .execute_operation(connection, request)
        .await
}

pub async fn plan_data_edit(
    connection: &ResolvedConnectionProfile,
    request: &DataEditPlanRequest,
) -> Result<DataEditPlanResponse, CommandError> {
    adapter_for_engine(&connection.engine)?
        .plan_data_edit(connection, request)
        .await
}

pub async fn execute_data_edit(
    connection: &ResolvedConnectionProfile,
    request: &DataEditExecutionRequest,
) -> Result<DataEditExecutionResponse, CommandError> {
    adapter_for_engine(&connection.engine)?
        .execute_data_edit(connection, request)
        .await
}

pub async fn inspect_permissions(
    connection: &ResolvedConnectionProfile,
) -> Result<PermissionInspection, CommandError> {
    adapter_for_engine(&connection.engine)?
        .inspect_permissions(connection)
        .await
}

pub async fn collect_diagnostics(
    connection: &ResolvedConnectionProfile,
    scope: Option<&str>,
) -> Result<AdapterDiagnostics, CommandError> {
    adapter_for_engine(&connection.engine)?
        .collect_diagnostics(connection, scope)
        .await
}
